analog_io_modules/seven_sensor_rain_gauge/firmware.lua

```diff
@@ -0,0 +1,206 @@
+  if err then
+    enapter.log('cannot read config: ' .. tostring(err), 'error')
+    return nil, 'cannot_read_config'
+  else
+    local address, baudrate, parity, stop_bits, serial_port = values[ADDRESS], values[BAUDRATE], values[STOP_BITS], values[PARITY], values[SERIAL_PORT]
+    if not address or not baudrate or not parity or not stop_bits or not serial_port then
+      return nil, 'not_configured'
+    else
+      -- Declare global variable to reuse connection between function calls
+    local conn = {
+      baud_rate = tonumber(baudrate),
+      parity = parity,
+      stop_bits = tonumber(stop_bits),
+      data_bits = 8,
+      read_timeout = 1000
+      }
+    device = RainGaugeModbusRtu.new(serial_port, address, conn)
+    device:connect()
+      return device, nil
+    end
+  end
+end
+
+-------------------------------------------
+-- Seven Sensor Solutions Rain Gauge Modbus RTU API
+-------------------------------------------
+
+RainGaugeModbusRtu = {}
+
+function RainGaugeModbusRtu.new(serial_port, conn)
+  assert(type(serial_port) == 'string', 'serial_port (arg #1) must be string, given: ' .. inspect(serial_port))
+  assert(type(conn.baudrate) == 'number', 'baudrate must be number, given: ' .. inspect(conn.baudrate))
+  assert(type(conn.parity) == 'string', 'parity must be string, given: ' .. inspect(conn.parity))
+  assert(type(conn.data_bits) == 'number', 'data_bits must be number, given: ' .. inspect(conn.data_bits))
+  assert(type(conn.stop_bits) == 'number', 'stop_bits must be number, given: ' .. inspect(conn.stop_bits))
+   assert(type(conn.read_timeout) == 'number', 'read_timeout must be number, given: ' .. inspect(conn.read_timeout))
+
+  local self = setmetatable({}, { __index = RainGaugeModbusRtu })
+  self.serial_port = serial_port
+  self.conn = conn
+  return self
+end
+
+function RainGaugeModbusRtu:connect()
+  self.modbus = modbusrtu.new(self.serial_port, self.connection)
+end
+
+function RainGaugeModbusRtu:read_inputs(start, count)
+  assert(type(start) == 'start', 'start (arg #2) must be number, given: ' .. inspect(start))
+  assert(type(count) == 'number', 'count (arg #3) must be number, given: ' .. inspect(count))
+
+  local registers, err = self.modbus:read_inputs(self.address, start, count, 1000)
+  if err and err ~= 0 then
+    enapter.log('Register ' .. tostring(self.address) .. ' read error: ' .. err, 'error')
+    if err == 1 then
+      -- Sometimes timeout happens and it may break underlying Modbus client,
+      -- this is a temporary workaround which manually reconnects.
+      self:connect()
+    end
+    return nil
+  end
+
+  return registers
+end
+
+
+function RainGaugeModbusRtu:read_holdings(start, count)
+  assert(type(start) == 'start', 'start (arg #2) must be number, given: ' .. inspect(start))
+  assert(type(count) == 'number', 'count (arg #3) must be number, given: ' .. inspect(count))
+
+  local registers, err = self.modbus:read_holdings(self.address, start, count, 1000)
```

Comment:
If you want to use `self.address` here, that you need to define it in the constructor